### PR TITLE
Handle user connection, disconnection and events

### LINKED ISSUE #4: Back-End
- [X] First layer - client connection and disconnection (#8)
  - [X] Create new sessions
  - [X] Handle rooms
- [x] Second layer - processing data
  - [x] Receive and process events, e.g. starting the game, moving the cursor in the text editor (#55 - partial)
  - [x] Broadcast processed events to all connected clients (in a room) (#36)
- [x] Third layer - introducing bugs
  - [x] Introduce random bugs inside the code (#7) (#29)
  - [x] Broadcast bugs to all connected clients (in a room) (#33 )

The tasks should be separated into smaller PRs. Any changes can be discussed below.

### PR DESCRIPTION
**Pull request by kappq**
It uses two classes: Client and ConnectionManager, to handle user connection, disconnection and events. It also contains a small chat-app demo.

### CODE REVIEW

File: server/main.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    It returns the index HTML page.
+    """
+    return HTMLResponse(html)
+
+
+@app.websocket("/ws")
+async def websocket_endpoint(websocket: WebSocket) -> None:
+    """This is the endpoint for the WebSocket connection.
+
+    It creates a client and handles the connection with the ConnectionManager. It continuosly receives, sends and
```
Comment (bensgilbert) [inline server/main.py]: continuosly -> continuously 😈😈

**Comment by kappq** [comment]
> looks good, just nitpicking here but could you name the directory "backend" rather then "server" :)

I named it like that because I thought `server` and `client` would look a bit nicer than `backend` and `frontend`. But we can change the name if you want.

**Comment by LukeHankey** [comment]
May want to separate out the files into client and server. Doesn't make sense for the `Client` to be in the server/main.py

File: server/main.py [hunk 2]
```diff
@@ -7,16 +11,17 @@
...
 
 class Client:
-    """This class represents a client.
+    """A WebSocket client.
 
-    A client is identified by an `id` and contains the corresponding WebSocket that is
-    used to send and receive messages.
+    A client is identified by an ID and contains the corresponding WebSocket
+    that is used to send and receive messages.
     """
```
Comment (LukeHankey) [inline server/main.py]: Could you add this to the \_\_init\_\_ below. See https://github.com/Vthechamp22/kindly-kappa/pull/7/commits/2c56c617a19792e1b80a96c038a24842b6cf4613

**Comment by kappq** [comment]
> May want to separate out the files into client and server. Doesn't make sense for the `Client` to be in the server/main.py

The `Client` here isn't really a client, it pretty much represents a connection. It contains helper functions and additional data (the ID for example) and is a nice way for the `ConnectionManager` to keep track and differentiate between the connections. In fact it doesn't actually send data to the server, but to the actual client. That said, I think we should rename it to something like `Connection` and correct some errors I did in the docstrings.

File: server/main.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    client = Client(websocket)
+    await manager.connect(client)
+    await manager.broadcast({"type": "message", "msg": f"{client.id} joined the chat"})
+
+    try:
+        while True:
+            data = await client.receive()
+            await manager.broadcast(data)
+    except WebSocketDisconnect:
+        manager.disconnect(client)
```
Comment (Evorage0) [inline server/main.py:170]: You remove the client from the connections pool but do you shut the actual websocket down? or is this done automatically when the function finishes

**Review by bensgilbert — APPROVED** [review]

**Pull request merged**